whyis/database/elasticstore.py

```diff
@@ -0,0 +1,469 @@
+        print("Response:",r.status_code, r.content)
+
+    def elastic_query(self, query):
+        query = {
+            "query": {
+                "nested" : {
+                    "path" : "graphs.@graph",
+                    "score_mode" : "avg",
+                    "query" : query
+                }
+            }
+        }
+
+        self.session.post(self.url+"/_refresh") # Ensure store is up-to-date before reading
+
+        response = self.session.post(self.url+"/_search",data=json.dumps(query))
+
+        return response.json()
+
+    def subgraph(self, query):
+        """\
+        Returns a graph containing all triples matching a given Elasticsearch query.
+        """
+
+        json_response = self.elastic_query(query)
+        g = rdflib.ConjunctiveGraph()
+
+        g.addN(json_ld_triples(json_response))
+        return g
+
+    def triples(self, spo=(None,None,None), context=None, txn=None):
+        """Returns a generator over all the triples matching the specified pattern."""
+        assert self.__open, "The Store must be open."
+
+        subject, predicate, object = spo
+
+        if context is not None and not isinstance(context, URIRef):
+            if context == self or isinstance(context, BNode):
+                context = None
+            else:
+                context = context.identifier
+                
+        if context == self or isinstance(context, BNode):
+            context = None
+        query = {
+            "bool" : {
+                "must" : [
+                ]
+            }
+        }
+        match_all = True
+        if subject is not None:
+            match_all = False
+            query['bool']['must'].append( { "term" : { "graphs.@graph.@id" : str(subject) } } )
+        if predicate is not None and object is None:
+            match_all = False
+            query['bool']['must'].append( { "exists" : { 'field': "graphs.@graph."+str(predicate) } } )
+        if object is not None:
+            match_all = False
+            obj_path = ''
+            if isinstance(object, rdflib.Literal):
+                obj_path = obj_path + '.@value'
+            else:
+                obj_path = obj_path + '.@id'
+            if predicate is not None:
+                query['bool']['must'].append( { "term" : {
+                      "graphs.@graph."+str(predicate) +obj_path : str(object)
+                    } } )
+            else:
+                query['bool']['must'].append( { "term" : {
+                     "graphs.@graph.@object" +obj_path : str(object)
+                    } } )
+        if context is not None:
+            match_all = False
+            query['bool']['must'].append( { "term" : {
+                 "graphs.@id" : str(context)
+                } } )
+
+        if match_all:
+            query = {'match_all': {}}
+
+        #print json.dumps(query)
+        json_response = self.elastic_query(query)
```

Comment:
Private

Reply:
What should be private, elastic_query?

Reply:
Yes, the store should encapsulate access to ES.